lisa/tools/hugepages.py

```diff
@@ -0,0 +1,115 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT license.
+import re
+from enum import Enum
+from typing import Any, Optional, Set
+
+from lisa.executable import Tool
+from lisa.tools.echo import Echo
+from lisa.tools.free import Free
+from lisa.tools.ls import Ls
+from lisa.tools.lscpu import Lscpu
+from lisa.tools.mkfs import FileSystem
+from lisa.tools.mount import Mount
+from lisa.util import LisaException, SkippedException
+
+PATTERN_HUGEPAGE = re.compile(
+    r"^.*hugepages-(?P<hugepage_size_in_kb>\d+)kB.*",
+)
+
+
+class HugePageSize(Enum):
+    HUGE_1GB = 1048576
+    HUGE_2MB = 2048
+    HUGE_512MB = 524288
+    HUGE_16GB = 16777216
+
+
+class Hugepages(Tool):
+    @property
+    def command(self) -> str:
+        return ""
+
+    @property
+    def can_install(self) -> bool:
+        return False
+
+    def _initialize(self, *args: Any, **kwargs: Any) -> None:
+        self._hugepage_sizes: Set[int] = set()
+        self._hugepages_dir_path = "/sys/devices/system/node/node*/hugepages"
+
+    def get_hugepage_sizes_in_kb(self) -> None:
+        """
+        This function get the hugepage sizes in kB available on the system.
+        This is done by listing the hugepages directory
+        and extracting the size from the directory name.
+        e.g. [
+        "'/sys/devices/system/node/node0/hugepages/hugepages-16777216kB/'",
+        "'/sys/devices/system/node/node0/hugepages/hugepages-2048kB/'",
+        "'/sys/devices/system/node/node0/hugepages/hugepages-524288kB/'"
+        ]
+        """
+        ls = self.node.tools[Ls]
+        # if not ls.path_exists(self._hugepages_dir_path, sudo=True):
+        #     raise SkippedException(f"path: {self._hugepages_dir_path} does not exist")
+        hugepage_size_dir_names = ls.list_dir(self._hugepages_dir_path, sudo=True)
+        for hugepage_size_dir_name in hugepage_size_dir_names:
+            matched_hugepage = PATTERN_HUGEPAGE.match(hugepage_size_dir_name)
+            if not matched_hugepage:
+                raise LisaException(
+                    f"no hugepage size found in {hugepage_size_dir_name}"
+                )
+            self._hugepage_sizes.add(int(matched_hugepage.group("hugepage_size_in_kb")))
+
+    def _enable_hugepages(
+        self, hugepage_size_kb: HugePageSize = HugePageSize.HUGE_2MB
+    ) -> None:
+        echo = self.node.tools[Echo]
+        meminfo = self.node.tools[Free]
+        nics_count = len(self.node.nics.get_nic_names())
+        numa_nodes = self.node.tools[Lscpu].get_numa_node_count()
+
+        request_space_kb = (nics_count - 1) * 1024 * 1024 * numa_nodes * 2
+        free_memory_kb = meminfo.get_free_memory_kb()
+
+        if free_memory_kb < request_space_kb:
+            self.node.log.debug(
```

Comment:
If there isn't enough space the allocation will probably fail and the test results will be useless; maybe you should raise an exception here. NotEnoughMemoryException or something (not sure if that type exists yet)